core/state/database.go

```diff
@@ -307,26 +324,28 @@ func (fdb *ForkingDB) EndTransition() {
 	  |____|  |___|  /\___        \___  |____/\___  |   __/|___|  (____  |___|  |__|      |___|  (____  /_____/     |____(____  |___|  \____ |\___  \____ |
                                                     |__|`)
 	fdb.ended = true
+	if err := fdb.VerkleDB.diskdb.Put(dbKeyOverlayTransitionEnded, []byte{1}); err != nil {
+		return fmt.Errorf("failed to mark transition as ended: %s", err)
+	}
+	return nil
 }
 
 func (fdb *ForkingDB) AddTranslation(orig, trans common.Hash) {
 	// TODO make this persistent
 	fdb.translatedRootsLock.Lock()
 	defer fdb.translatedRootsLock.Unlock()
-	fdb.translatedRoots[fdb.translationIndex] = trans
-	fdb.origRoots[fdb.translationIndex] = orig
-	fdb.translationIndex = (fdb.translationIndex + 1) % len(fdb.translatedRoots)
+	fdb.diskdb.Put(append(translatedRootPrefix, orig.Bytes()...), trans.Bytes())
+	log.Info("Saved translation", "originRoot", orig.Bytes(), "translatedRoot", trans.Bytes())
 }
 
 func (fdb *ForkingDB) getTranslation(orig common.Hash) common.Hash {
 	fdb.translatedRootsLock.RLock()
 	defer fdb.translatedRootsLock.RUnlock()
-	for i, o := range fdb.origRoots {
-		if o == orig {
-			return fdb.translatedRoots[i]
-		}
+	trans, err := fdb.diskdb.Get(append(translatedRootPrefix, orig.Bytes()...))
+	if err != nil {
+		return common.Hash{}
 	}
```

Comment:
Technically speaking, this is equivalent to before. If `Get` returns `err=leveldb.NotFound` it returns `common.Hash{}` as before. 

There's an extra border-case here if the `err` is related to a disk error. Very low chance for now to care about, but that needs bubbling some error in the chain. Not sure if this code will stay for long, but we can properly handle and clean up after.